### ISSUE TITLE
Dagger linter warning about using @JvmStatic annotation

### ISSUE THREAD

**Issue by bitvale** [google/dagger#3095]
`DaggerKotlinIssueDetector` has a warning: "It's redundant to annotate @Provides functions in object classes with @JvmStatic".

By I think it is redundant for `companion object` and not for `object` declaration. As an example, for this is Kotlin module:

![image](https://user-images.githubusercontent.com/14936809/144874747-960e0105-55c8-4668-9dad-73a04b47ebd7.png)

Dagger generates this code:

![image](https://user-images.githubusercontent.com/14936809/144875087-d8961b2c-7532-407f-aaae-31973b1089fa.png)

As you can see, Dagger use `INSTANCE` field to provide a dependency.

And this is the same, but with `@JvmStatic` annotation:

![image](https://user-images.githubusercontent.com/14936809/144875134-69989289-8ca3-4f47-93e8-fc59a1030537.png)

And generated code:

![image](https://user-images.githubusercontent.com/14936809/144875362-4bbbc8fb-82f7-4de7-ba47-df5c63609e8f.png)

In this case Dagger provide dependency directly: `SomeDataModule.provideApiService(retrofit)`

But there is a warning:

![image](https://user-images.githubusercontent.com/14936809/144875648-d9b844f3-11c4-43e1-875b-63fa3b0661c4.png)

It is true for `companion object`, because there is no sense if we use `@JvmStatic` annotation or not, Kotlin each time use `INSTANCE` to call a method or access a field form a `companion object`, and with `@JvmStatic` it will simply generate one more static method that proxy the same call to `INSTANCE` field.

**Comment by bcorso** [comment]
Hi @bitvale,

Is there a particular reason that you want to use `@JvmStatic`?

You're right that the generated code isn't exactly the same in both cases, but functionally it should be pretty close. The difference is essentially whether we call 

  1) a static method, or
  2) a non-static method of a static instance

This can be seen from the stubs below:

```
// Java stub with @JvmStatic
@Module
public final class MyModule {
    @JvmStatic
    @Provides
    public static final SomeApiService provideSomeApiService(...) { ... }
}

// Java stub without @JvmStatic
@Module
public final class MyModule {
    // Note: this object will only be created once per process.
    public static final MyModule INSTANCE = null;

    @Provides
    public final SomeApiService provideSomeApiService(...) { ... }
}
```

The second case is more Kotlin friendly, as `object` methods are meant to replace static methods in Kotlin. This is why the linter prefers that you avoid using `@JvmStatic. However, we'd be interested to hear if there are cases where `@JvmStatic` is still preferred.

**Comment by bitvale** [comment]
You are right about "functionally it should be pretty close". But I don't need instance access to call the static method)

For example, Android R8 optimization removes it. Jake Wharton has written about this optimization in [this article](https://jakewharton.com/r8-optimization-staticization/#more-than-companions.).

I'm not sure if this has any significant effect on performance, although somewhere it might be. But the generated code is different (with or without annotation) and the linter tells me that it is not friendly and redundant. But why the R8 tools will optimize it?) What if I don't use R8 and want to call method directly without `INSTANCE`?

**Comment by bcorso** [comment]
> the linter tells me that it is not friendly and redundant

I think our assumption is that most would prefer not to have annotations like `@JvmStatic` throughout their code if they don't need it. 

Although, in your case it sounds like you're saying you need it because you are calling the `@Provides` directly? I'm curious why you're doing this though, is it for testing?

In any case, I  if you _really_ don't agree with this check, I think you can disable it for your library/project (https://developer.android.com/studio/write/lint#gradle), right?

Also, the history behind this is that `@JvmStatic` _used to be_ required, but we added special handling in Dagger to avoid the need for `@JvmStatic`, and this linter is a hint to people that it's no longer needed.

> What if I don't use R8?

Hmm, if you're not using R8 then you likely have bigger issues since none of your `companion`/`object` methods will be optimized, not just Dagger's.

**Comment by bitvale** [comment]
> Although, in your case it sounds like you're saying you need it because you are calling the @Provides directly? I'm curious why you're doing this though, is it for testing?

Dagger generated code (that is generated with `@JvmStatic`) is calling the @Provides directly). In my first post there are two pictures with Dagger generated code with and without annotation. Or not? I just explore Dagger generated code)

> In any case, I if you really don't agree with this check, I think you can disable it for your library/project (https://developer.android.com/studio/write/lint#gradle), right?

Of course

> Also, the history behind this is that @JvmStatic used to

It became clearer about the history)

> Hmm, if you're not using R8 then you likely have bigger issues since none of your companion/object methods will be optimized, not just Dagger's.

It's true, but the issue is about Dagger optimization)


Thank you for your time and explanation! I think the issue has been resolved.

**Comment by bcorso** [comment]
Would it help if we correct the wording? I think you're right that the use of "redundant" is not really correct, so changing it to something more accurate my help.

I also don't think we're opposed to removing this lint check if it seems like a majority of people find it unhelpful/annoying, but I'm not sure we're there yet.

**Comment by bitvale** [comment]
> Would it help if we correct the wording? I think you're right that the use of "redundant" is not really correct, so changing it to something more accurate my help.

In my opinion, the code that is generated by Dagger with `@JvmStatic` and without is different and with annotation is slightly more optimized as we discussed above. From this point of view, the annotation looks strange. It is not very clear what text can be used to replace the warning: "It's redundant but slightly more optimized":)

As about `companion object` I agree at all, it's redundant, because although the generated code is different, but without annotation we omit the call of an extra static method. And the warning tell us that it is redundant.

> a majority of people find it unhelpful/annoying

A majority of people don't think about it at all)
